shotover/src/codec/opensearch.rs

```diff
@@ -0,0 +1,323 @@
+use super::{CodecBuilder, CodecReadError, CodecWriteError, Direction};
+use crate::frame::{opensearch::HttpHead, Frame, MessageType, OpenSearchFrame};
+use crate::message::{Encodable, Message, Messages};
+use anyhow::Result;
+use bytes::{Buf, BytesMut};
+use http::{header, HeaderName, HeaderValue, Request, Response};
+use std::sync::Arc;
+use tokio_util::codec::{Decoder, Encoder};
+
+#[derive(Clone)]
+pub struct OpenSearchCodecBuilder {
+    direction: Direction,
+}
+
+impl CodecBuilder for OpenSearchCodecBuilder {
+    type Decoder = OpenSearchDecoder;
+    type Encoder = OpenSearchEncoder;
+
+    fn new(direction: Direction) -> Self {
+        Self { direction }
+    }
+
+    fn build(&self) -> (OpenSearchDecoder, OpenSearchEncoder) {
+        (
+            OpenSearchDecoder::new(self.direction),
+            OpenSearchEncoder::new(self.direction),
+        )
+    }
+
+    fn websocket_subprotocol(&self) -> &'static str {
+        "opensearch" //TODO
+    }
+}
+
+pub struct OpenSearchDecoder {
+    direction: Direction,
+    state: State,
+}
+
+impl OpenSearchDecoder {
+    pub fn new(direction: Direction) -> Self {
+        Self {
+            direction,
+            state: State::ParsingResponse,
+        }
+    }
+
+    fn decode_request(&self, src: &mut BytesMut) -> Result<Option<(usize, HttpHead, usize)>> {
+        let mut headers = [httparse::EMPTY_HEADER; 16];
+        let mut request = httparse::Request::new(&mut headers);
+
+        let amt = match request.parse(&src[0..src.len()]).unwrap() {
+            httparse::Status::Complete(amt) => amt,
+            httparse::Status::Partial => return Ok(None),
+        };
+        match request.version.unwrap() {
+            1 => (),
+            _version => panic!("error!"), // TODO version error
+        }
+
+        let mut builder = Request::builder();
+
+        builder = builder
+            .method(request.method.unwrap().to_owned().as_str())
+            .uri(request.path.unwrap().to_owned().as_str());
+
+        let builder_headers = builder.headers_mut().unwrap();
+
+        let request_headers: Vec<(String, Vec<u8>)> = request
+            .headers
+            .iter()
+            .take_while(|h| !h.name.is_empty() && !h.value.is_empty())
+            .map(|h| (h.name.to_owned(), h.value.to_owned()))
+            .collect();
+
+        for (name, value) in request_headers {
```

Comment:
These can be combined to avoid reallocating all of those strings.
```rust
        for header in request.headers {
            if header.name.is_empty() && header.value.is_empty() {
                break;
            }
            builder_headers.insert(
                HeaderName::from_bytes(header.name.as_bytes()).unwrap(),
                HeaderValue::from_bytes(header.value).unwrap(),
            );
        }
```